src/Microsoft.AspNetCore.Server.HttpSys/RequestProcessing/RequestStream.cs

```diff
@@ -418,6 +425,40 @@ public override void EndWrite(IAsyncResult asyncResult)
             throw new InvalidOperationException(Resources.Exception_ReadOnlyStream);
         }
 
+        // Called before each read
+        private void CheckSizeLimit()
+        {
+            // Note SwitchToOpaqueMode sets HasStarted and clears _maxSize, so these limits don't apply.
+            if (!HasStarted)
+            {
+                var contentLength = RequestContext.Request.ContentLength;
+                if (contentLength.HasValue && _maxSize.HasValue && contentLength.Value > _maxSize.Value)
+                {
+                    throw new IOException(
+                        $"The request's Content-Length {contentLength.Value} is larger than the request body size limit {_maxSize.Value}.");
+                }
+
+                HasStarted = true;
+            }
+            else if (TryCheckSizeLimit(0, out var exception))
```

Comment:
This is needed for the read-after-exception scenario. E.g. The first read went to wire, got data, incremented _totalRead past the limit, and threw. If the caller ignores that exception and calls read again then this does the check again and throws the same error again.